[content-visibility] Do no create renderers for skipped content

rwlbuis opened the pull request:
#### d73b0dac4dd6f8ef37e56fecf5a7ef0abc5a0087
<pre>
[content-visibility] Do no create renderers for skipped content
<a href="https://bugs.webkit.org/show_bug.cgi?id=252658">https://bugs.webkit.org/show_bug.cgi?id=252658</a>

Reviewed by NOBODY (OOPS!).

Do no create renderers for skipped content unless we are using a DOM method
that measures layout, in this case we use ContentVisibilityForceLayoutScope.

Since in some places in DOM we need to check for skipped content, in the case
where there is no renderer store the style using storeSkippedContentStyle and
use that when isSkippedContent is called while there is no renderer.

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/css/css-contain/content-visibility/content-visibility-035.html
* LayoutTests/imported/w3c/web-platform-tests/css/css-contain/content-visibility/content-visibility-044-expected.txt:
* Source/WebCore/dom/Document.cpp: Add helper to force layout if there is a hidden content-visibility subtree.
(WebCore::ContentVisibilityForceLayoutScope::ContentVisibilityForceLayoutScope):
(WebCore::ContentVisibilityForceLayoutScope::~ContentVisibilityForceLayoutScope):
* Source/WebCore/dom/Document.h: keep track of whether we want to ignore skipped content.
(WebCore::Document::ignoreSkippedContent):
(WebCore::Document::isSkippedContentIgnored const):
* Source/WebCore/dom/Element.cpp:
(WebCore::Element::hasFocusableStyle const):
(WebCore::Element::isFocusable const): rely on Element.isSkippedContent.
(WebCore::Element::offsetLeftForBindings):
(WebCore::Element::offsetTopForBindings):
(WebCore::Element::scrollTop):
(WebCore::Element::getClientRects):
(WebCore::Element::boundingClientRect):
(WebCore::Element::isSkippedContent const): check for skipped content regardless whether there is a renderer.
(WebCore::Element::storeSkippedContentStyle):
(WebCore::Element::innerText): rely on Element.isSkippedContent.
* Source/WebCore/dom/Element.h:
* Source/WebCore/html/HTMLImageElement.cpp:
(WebCore::HTMLImageElement::width):
(WebCore::HTMLImageElement::height):
(WebCore::HTMLImageElement::x const):
(WebCore::HTMLImageElement::y const):
* Source/WebCore/html/ImageInputType.cpp:
(WebCore::ImageInputType::height const):
(WebCore::ImageInputType::width const):
* Source/WebCore/page/ResizeObservation.cpp:
(WebCore::ResizeObservation::computeObservedSizes const):
* Source/WebCore/rendering/RenderObject.cpp:
(WebCore::RenderObject::shouldSkipContent const): only can apply if size containment applies
* Source/WebCore/rendering/RenderObject.h: move from RenderElement
* Source/WebCore/rendering/RenderObjectInlines.h:
(WebCore::RenderObject::shouldApplySizeOrStyleContainment const):
* Source/WebCore/rendering/updating/RenderTreeUpdater.cpp:
(WebCore::RenderTreeUpdater::updateRenderTree): always proceed processing skipped content to set style on element.
(WebCore::RenderTreeUpdater::updateElementRenderer): do not create renderers for skipped content.
(WebCore::RenderTreeUpdater::updateTextRenderer): do not create text renderers for skipped content.
* Source/WebCore/style/StyleChange.cpp:
(WebCore::Style::determineChange): a content-visibility change needs a renderer update.
* Source/WebCore/svg/SVGLocatable.cpp:
(WebCore::SVGLocatable::getBBox):
* Source/WebCore/svg/SVGTextContentElement.cpp:
(WebCore::SVGTextContentElement::getNumberOfChars):
(WebCore::SVGTextContentElement::getComputedTextLength):
(WebCore::SVGTextContentElement::getCharNumAtPosition):
</pre><!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d73b0dac4dd6f8ef37e56fecf5a7ef0abc5a0087

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/5772 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/5935 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/6124 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/7329 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/6164 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/5772 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/6157 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/5903 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/7938 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/5877 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/5924 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/5216 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/7383 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/3401 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/5200 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/12983 "4 flakes 158 failures") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/5269 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/5277 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/7440 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/5721 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/4694 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/5166 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/5130 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/9283 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/5527 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

mrobinson commented on Source/WebCore/rendering/updating/RenderTreeUpdater.cpp:
```diff
@@ -344,19 +344,22 @@ void RenderTreeUpdater::updateElementRenderer(Element& element, const Style::Ele
+    else if (isSkippedContent)
+        element.storeSkippedContentStyle(makeUnique<RenderStyle>(WTFMove(elementUpdateStyle)));
     else
         element.clearDisplayContentsStyle();
```
I think it might be better to move the check for `hasDisplayContents` into the definition of `isSkippedContent` and include the specification language and link to the spec:

```c++
// The specification https://drafts.csswg.org/css-contain/#size-containment says that
// content-visibility has no effect on an element:
// "if the element does not generate a principal box (as is the case with display: contents"
```

In addition, should `element.clearDisplayContentsStyle()` be called here even if the element skips its content?

rwlbuis replied:
I think these are kind of independent things, i.e we need to keep track of both display:contents and is skipped content to make good decisions. The dependent thing is that we need to decide where the style goes to as we can only move it once. I decided to give display: contents priority for that, but it does mean I had to adapt `Element::isSkippedContent()` slightly.

I did miss the `element.clearDisplayContentsStyle()` thing initially, well spotted!

mrobinson commented on Source/WebCore/html/HTMLImageElement.cpp:
```diff
@@ -508,6 +508,8 @@ void HTMLImageElement::setPictureElement(HTMLPictureElement* pictureElement)
 unsigned HTMLImageElement::width()
 {
+    ContentVisibilityForceLayoutScope scope(*const_cast<HTMLImageElement*>(this));
+
```
In this method and all of the ones in `HTMLImageElement.cpp`, I think it would be good to create a block around the case where the renderer exists and to move the scope defined here into that block. That way we aren't doing extra work if the element has not been added to the document.

rwlbuis replied:
Looking at this again I think we can only make a decision about the renderer after recomputing style and layout, so I think the positioning is correct, apart from the inRenderedDocument blocks, which we discussed.

mrobinson commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -2294,6 +2294,26 @@ bool Document::updateStyleIfNeeded()
+    if (m_document)
+        m_document->ignoreSkippedContent(false);
+}
+
```
I think it's good to either:
- Handle the case where two scopes are created in a recursive manner and avoid unignoring skipped content.
- Add an assertion that shows that scopes aren't created recursively.

rwlbuis replied:
I went for the second option.

mrobinson commented on LayoutTests/TestExpectations:
```diff
@@ -4828,13 +4828,14 @@ webanimations/translate-property-and-translate-animation-with-delay-on-forced-la
-imported/w3c/web-platform-tests/css/css-contain/content-visibility/content-visibility-035.html [ Skip ]
+# can't work since scroll offset is stored in renderer
+imported/w3c/web-platform-tests/css/css-contain/content-visibility/content-visibility-081.html [ Skip ]
+# problematic test crbug.com/1247844
```
Nit: You might want to write a bit more here about why this test is failing on WebKit. That is, comment on the fact that this is due to scroll offsets being stored in renderers.

rwlbuis replied:
I think the comment covers it, or do you mean write more about t in the commit log?

mrobinson reviewed (changes requested):
This is looking pretty good to me, but I have a few minor comments.

rwlbuis committed:
[content-visibility] Do no create renderers for skipped content
https://bugs.webkit.org/show_bug.cgi?id=252658

Reviewed by NOBODY (OOPS!).

Do no create renderers for skipped content unless we are using a DOM method
that measures layout, in this case we use ContentVisibilityForceLayoutScope.

Since in some places in DOM we need to check for skipped content, in the case
where there is no renderer store the style using storeSkippedContentStyle and
use that when isSkippedContent is called while there is no renderer.

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/css/css-contain/content-visibility/content-visibility-035.html
* LayoutTests/imported/w3c/web-platform-tests/css/css-contain/content-visibility/content-visibility-044-expected.txt:
* Source/WebCore/dom/Document.cpp: Add helper to force layout if there is a hidden content-visibility subtree.
(WebCore::ContentVisibilityForceLayoutScope::ContentVisibilityForceLayoutScope):
(WebCore::ContentVisibilityForceLayoutScope::~ContentVisibilityForceLayoutScope):
* Source/WebCore/dom/Document.h: keep track of whether we want to ignore skipped content.
(WebCore::Document::ignoreSkippedContent):
(WebCore::Document::isSkippedContentIgnored const):
* Source/WebCore/dom/Element.cpp:
(WebCore::Element::hasFocusableStyle const):
(WebCore::Element::isFocusable const): rely on Element.isSkippedContent.
(WebCore::Element::offsetLeftForBindings):
(WebCore::Element::offsetTopForBindings):
(WebCore::Element::scrollTop):
(WebCore::Element::getClientRects):
(WebCore::Element::boundingClientRect):
(WebCore::Element::isSkippedContent const): check for skipped content regardless whether there is a renderer.
(WebCore::Element::storeSkippedContentStyle):
(WebCore::Element::innerText): rely on Element.isSkippedContent.
* Source/WebCore/dom/Element.h:
* Source/WebCore/html/HTMLImageElement.cpp:
(WebCore::HTMLImageElement::width):
(WebCore::HTMLImageElement::height):
(WebCore::HTMLImageElement::x const):
(WebCore::HTMLImageElement::y const):
* Source/WebCore/html/ImageInputType.cpp:
(WebCore::ImageInputType::height const):
(WebCore::ImageInputType::width const):
* Source/WebCore/page/ResizeObservation.cpp:
(WebCore::ResizeObservation::computeObservedSizes const):
* Source/WebCore/rendering/RenderObject.cpp:
(WebCore::RenderObject::shouldSkipContent const): only can apply if size containment applies
* Source/WebCore/rendering/RenderObject.h: move from RenderElement
* Source/WebCore/rendering/RenderObjectInlines.h:
(WebCore::RenderObject::shouldApplySizeOrStyleContainment const):
* Source/WebCore/rendering/updating/RenderTreeUpdater.cpp:
(WebCore::RenderTreeUpdater::updateRenderTree): always proceed processing skipped content to set style on element.
(WebCore::RenderTreeUpdater::updateElementRenderer): do not create renderers for skipped content.
(WebCore::RenderTreeUpdater::updateTextRenderer): do not create text renderers for skipped content.
* Source/WebCore/style/StyleChange.cpp:
(WebCore::Style::determineChange): a content-visibility change needs a renderer update.
* Source/WebCore/svg/SVGLocatable.cpp:
(WebCore::SVGLocatable::getBBox):
* Source/WebCore/svg/SVGTextContentElement.cpp:
(WebCore::SVGTextContentElement::getNumberOfChars):
(WebCore::SVGTextContentElement::getComputedTextLength):
(WebCore::SVGTextContentElement::getCharNumAtPosition):
